### PR TITLE
Fix issue for systemless root with no free space

### LINKED ISSUE #781: AdAway incompatible with newer systemless root
Even after applying the "unofficial" systemless root Zip from [XDA](http://forum.xda-developers.com/showthread.php?t=2190753), AdAway does not work properly with systemless root—which is probably the most popular root method on XDA by this point.

If "Target hosts file" is set to "/system/etc/hosts" (default), AdAway will always fail with a message that it's out of space on /system for a Nexus 5X or 6 running Marshmallow—most devices running Marshmallow, apparently—even though we're working with a symlink on an overlay. If there's a way to prevent the freespace check, I can't find it.

There's one way AdAway can be used with systemless root, but only if:

- "Target hosts file" is set to "Custom"
- "Custom target" is set to "/su/etc/hosts"
- You don't mind responding to a message about "creating a symlink" on every hosts file update (automatic updates break completely)

Please add proper support for symlinking - or at least provide a way to disable freespace and file creation checks. Without systemless root, I have to reflash the entire /system partition every month as the OTAs roll in. (Nexus 6.)

Thanks!

### PR DESCRIPTION
**Pull request by PerfectSlayer**
Systemless root is now the default root behavior.
Some devices has no free space left on /system partition (like Nexus 6).
AdAway is raising a NotEnoughSpaceException whereas /system/etc/hosts is
mount -bind on /su partition where space left.

Should fix #781 issue.
Few more details in [this xda post](http://forum.xda-developers.com/showpost.php?p=67817379&postcount=6554).

### CODE REVIEW

**Comment by 0-kaladin** [comment]
I had compiled an adaway in the xda thread that did your change.  I didn't merge it cause the real fix I want to implement is to detect if systemless root is installed (or even have the user simply hit a checkbox stating as such), then the free disk space can check the /data partition if that is true.  that would be the real fix.  this is simply using a different freespace method that will look at space reserved by the system and so doesn't see /system as being full (even though it is to the user).

**Comment by PerfectSlayer** [comment]
I'm still looking for a way to check systemless root… Maybe check and parse the su version? (with ```su -V```)

Otherwise, we could check if the ```/su/etc/hosts``` file is present so systemless root and AdAway systemless script are installed. So, we can assure ```/data``` is used and check free space on this partition instead of ```/system```.

@0-kaladin What do you think about it?

**Comment by 0-kaladin** [comment]
Honestly, to start i wouldn't even check automatically.  I'd have a check box for the user to check if they want the systemless functionality or not.  Because they might be systemless root, yet have something like elementalX kernel and hence adaway allowed to write to system and maybe they want that.  So having a manual checkbox I think is something that should be there regardless.  if they check the box for systemless, then it should do the things the flashable zip does in the Adaway XDA thread and prompt them to reboot.  
Once this is working as intended we can add an auto detection of root method which would effect the same checkbox the user can manually override.

**Comment by mwyau** [comment]
@0-kaladin I agree, that's what I am thinking too. How about adding a check box for systemless, which disables warning of creating the symbolic link, and add "/magisk/.core/hosts" and "/su/etc/hosts" to the custom target pull down list?  It's safer than writing on the bind mounted /system/etc/hosts since if somehow it's not bind mounted, the actual system file will be overwritten.

**Comment by PerfectSlayer** [comment]
Ok, so I propose to edit my PR to:
- Add a checkbox in settings for systemless settings,
- Add ```/magisk/.core/hosts``` and ```/su/etc/hosts``` to apply method entries.

Checking the systemless option will:

1. Copy the current apply method entry location file to ```/su/etc/hosts```,
2. Write the ```0000adaway.script``` to ```/su/su.d/```,
3. (Change current value of apply method entry to ```/su/etc/hosts```?)
4. Ask to reboot.

Unchecking the systemless option will:

1. Remove ```/su/su.d/0000adaway.script``` and ```/su/etc/hosts``` files
2. (Set default value to apply method entry settings?)

----
Otherwise, checking/unchecking the systemless checkbox could be simplify to write/remove the ```000adaway.script``` and ask the user to reapply hosts after reboot. Which solution do you prefer?

**Comment by 0-kaladin** [comment]
Ok so per your comments above
**Checking the systemless option will:**

1.     Copy the current apply method entry location file to /su/etc/hosts,
2.     Write the 0000adaway.script to /su/su.d/,
3.     (Change current value of apply method entry to /su/etc/hosts?)
4.     Ask to reboot.

I would not do your number 3.  the script in su/su.d/ will bind mount /su/etc/hosts to /system/etc/hosts.  so leaving adaway pointed to /system/etc/hosts will really be writing to /su/etc/hosts.  
Now per the free space issue, in this case since adaway *thinks* it is writing to /system yet with the bind mount it is writing to /su.  the superuser install created /su out of /data.  So my point is in a systemless install mode you should check free space but just always check /data for free space not /system.  So that is another step that should happen when systemless check box is checked.  Well not a step that should happen when the box is checked but that is what should happen ONCE the box has been checked and the user goes to apply adaway.  The check for free space should just check /data if the systemless option is selected.

**Comment by 0-kaladin** [comment]
BTW, next weekend i'll pull out a nexus 5 and compile all this and test it out. 
Thanks a ton for contributing, its great to get extra devs in here!

**Comment by PerfectSlayer** [comment]
You're welcome. I used this app from years now, so it's the least I can do :wink: 

I start adding support I found some ways to simplify the systemless support:
* Instead of copying the current apply method entry to ```/su/etc/hosts```, I copy the ```/system/etc/hosts```. If other location is used, symlink should do the trick. Ok with it?
* Is adding ```/magisk/.core/hosts``` and ```/su/etc/hosts``` to apply method entries really useful? Since the hosts file is mounted, it's pointless to select custom or specific apply method entry, isn't it?
* Does the free space on ```/data``` is the same of ```/su```? Even if ```/su``` is a loop device in ```/data/su.img```? I saw [Chainfire increase the image from 32mo to 96mo](http://forum.xda-developers.com/showpost.php?p=64916199&postcount=3) so I guess we should check ```/su``` partition instead of ```/data``` ?

Keep in mind I still developing (and testing) the feature. But you advices and feedback on how achieve it are really useful for me!

**Comment by 0-kaladin** [comment]
Yes ok with your first bullet point.
Is adding those other apply method's really useful?  I don't think so.  when in systemless mode you should be leaving the apply location to /system/etc/hosts  - still those who have a reason to go elsewhere, can use the custom option.  I fear if /su/etc/hosts is an option they may think they need to use it when systemless which is not true.
regarding /su and /data, yea since su is an image mounted on /data then might as well check /su for space.

**Comment by mwyau** [comment]
- Should we not hardcoding "2. Write the 0000adaway.script to /su/su.d/," and leave it to the recovery update package? The root environment changes all the time. More users are moving to Magisk and phh's SuperUser now. Are we going to change Adway every time a new rooting software emerges?
- By writing directly to /magisk/.core/hosts or /su/etc/hosts, I think it is a good safety precaution to prevent users modifying the /system partition by accident and fails OTA updates.

BTW, if I want to contribute here, should I fork from PerfectSlayer/AdAway and make a pull request there?

**Comment by PerfectSlayer** [comment]
@0-kaladin By adding ```/su/etc/hosts``` apply location, I know I must check /su partition free space (without hacking the checkFreeSpaceMethod). 

@mwyau Chainfire SuperSU seems stable to me. I don't know if it's the case for Magisk and phh's SuperUser. It's a global issue that all root methods. Should we support all? I mean hardcode all locations or leave it optional and add some complexity (means issues) to AdAway?

For the now, I'm trying to add the global systemless support. It will be made step by step (maybe just SuperSU support to start?) and I'm opened to debate. As you can see, I would have an open development approach by asking you how it should be done and regularly pushing my commits (and I hope I do not send you too much spam :fearful:)
It could be to early to contribute as the feature is still in early development (some commits are being reverts, I'm still looking how to achieve it, etc…) but if you would like to contribute, you could:
- Add comment to commit in the PR (for bug, improvements to do on the code),
- Discuss here on how the feature should work,
- And yeah, I think you could fork my master, make PR to it and then I have to approve them (As said, it could give some headache doing PR-ception as I still in early development :wink:).

**Comment by PerfectSlayer** [comment]
I'm reading [the developer documentation about Magisk](http://forum.xda-developers.com/showpost.php?p=68034110&postcount=4) and it's said ```/magisk/hosts``` will be mounted automatically.

So it seems the only setting you have to apply in order to get systemless AdAway working with Magisk is to set custom location to ```/magisk/hosts```. Free space check should work as expected, and mount will be done by Magisk.

@mwyau Could you confirm this behavior?

**Comment by PerfectSlayer** [comment]
@0-kaladin Do you know why are you checking free space of the parent folder of target?
It causes some issue while checking bind mounted file… (checking parent directory is now where file are stored).

I ran some tests with hosts file mounted on ```/su/etc/hosts```:

| Location | Statfs free space | Remark |
|-------------------|-------------:|---|
| /system/etc          | 293044224 | Check parent folder |
| /system/etc/hosts |   92053504 | Checking file itself |
| /system                |   293044224 |   |
| /su                       | 92053504 |   |

So it seems to works well to check the file itself. There is a limitation: we can't use Statfs on symbolic link but the app didn't do this (it checks destination, not the link it creates). So is it okay to you to check the file itseft instead? It could simplify apply method entry locations a lot!

**Comment by mwyau** [comment]
@PerfectSlayer `/magisk/.core/hosts` was created and mounted by Magisk already. Just changing the custom target works, but as I suggested before, there should be an option to disable the symbolic link prompt.

Current behavior is that: after an update to the custom target `/magisk/.core/hosts`, if I choose `No` to not create the symbolic link, AdAway will be showing `Disabled`. But if I quit the app and reopen, it will show `Enabled: Newest hosts files enabled`.

My suggestion:
- Add an option for create symbolic link for custom target, which is selected by default
- After an update, if custom target is set and create symbolic link option is not selected, only prompt for reboot
- If Hide reboot dialog is also selected, simply detect the hosts file and show `Newest hosts files enabled`

Alternatively, is it possible to detect if `/system/etc/hosts` is bind mounted without root shell? It could be simply detecting where it is bind mounted, and write to that target automatically.

**Comment by PerfectSlayer** [comment]
Actually, I don't understand why you are using custom target with Magisk. Since the file is bind mounted, you should stick with the default location and no symbolic link will be proposed to you.
Doesn't it work fine with default location?

**Comment by 0-kaladin** [comment]
yea @PerfectSlayer i agree about magisk, i don't see a reason to change location.  if bind mounted keep location default then.  Also yes good tests with statfs, check the file or rather whatever they user has their hosts file set to.  should work well.  
bottom line I wouldn't worry about magisk or the other root for now.  they should bind mount the hosts if systemless anyway then default location should work.

**Comment by mwyau** [comment]
@PerfectSlayer I used custom target with Magisk because Nexus 6's system partition of Android 6.0 does not have enough space. This is not an issue now since the Android 7.0 image of Nexus 6 has a few hundreds megabytes of space in the system partition.

If checking the file itself works, I still think there should be an option to disable "suggesting creation of symbolic link" if custom target is set.

**Comment by PerfectSlayer** [comment]
@mwyau I know, I also own a Nexus 6 😉 
The symlink option is not required. That's the beauty of the simplest solution! If you use default location setting, no symbolic link is created are the space is now well check (I mean, it check the space of the mounted file, not the space of the system file).

@0-kaladin Did you build and try this branch?

**Comment by voltaek** [comment]
Any recent progress on this @0-kaladin or @PerfectSlayer ? I'm still running 6.0 on my Nexus 6 so I'd be happy to run a test build of this to see if the 'no free space' issue is resolved. Currently using Magisk as well, but I can test it with and without.

**Comment by PerfectSlayer** [comment]
I add the missing reboot dialog and I also fixed an issue which mount /su partition as read-only (preventing to disable systemless mode after applying hosts file).
Everything should be done now..!

File: AdAway/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -105,6 +105,8 @@
     <string name="download_fail_title">Download failed</string>
     <string name="no_private_file">Private file can not be created.</string>
     <string name="no_private_file_title">Private file creation failed</string>
+    <string name="revert_systemless_successful">You disable the systemless mode.\nIt may be necessary to reboot Android for the changes to take effect.\n\nDo you want to reboot?\n(To prevent data loss make sure no app uses the sdcard at the moment!)</string>
```
Comment (voltaek) [inline AdAway/src/main/res/values/strings.xml]: "You have disabled systemless mode." or "You disabled systemless mode." are better for the first sentence.

**Comment by voltaek** [comment]
@PerfectSlayer Sure, if you can give me a download link for the test build APK I'd be happy to test it out.

File: AdAway/src/main/res/values/strings.xml (hunk 1 above)
Comment (PerfectSlayer) [inline AdAway/src/main/res/values/strings.xml]: Thanks! I fixed it (and also change string key to be more relevant to the related action)

**Comment by PerfectSlayer** [comment]
Hi @voltaek,
Here is a debug build file of the d25f4cc commit:  [AdAway-debug.zip](https://github.com/AdAway/AdAway/files/545784/AdAway-debug.zip)

**Comment by voltaek** [comment]
@PerfectSlayer I've done some testing tonight, and everything seems to be working as intended. Here is some feedback and my testing results:

### No Magisk
1. With systemless mode enabled, uses `/su/etc/hosts` (via `/system/etc/hosts`) and puts `/su/su.d/0000adaway.script` in place, as expected.
1. Without systemless mode enabled, `/system` is written to with no warning to user. In my case it didn't actually do it because there isn't enough space to, but it would have had there been space.
 * Now that systemless root is the default rooting method for most, I think it would be best if we show a warning about writing to `/system`.
   * Would only do so when systemless mode has not been enabled and the `hosts` file setting is under any `/system` path.
 * Can we keep a backup of the stock `hosts` file when writing anywhere in `/system`?
    * Doing this only matters if restoring it would restore the integrity of `/system` to a not-touched state for things like OTA and SafetyNet checks, and I'm not sure if putting the original file back would accomplish this or not.
    * Also we'd have to put it somewhere we know would always have disk space, since `/system` doesn't always have extra space.
    * The 'Disable Ad Blocking' button doesn't seem to set the file back to its original size or modified date, so I don't think this is already handling this scenario.
2. Can we alert the user to enable systemless mode when applicable?
 * I would hate to forget to check the option and overwrite `/system/etc/hosts`, though this goes back to my last point about warning the user about doing so.
 * When `su` is installed systemless, we could prompt the user to enable it.
 * Hopefully there is a simple way to check if one of the common systemless roots exists,

### With Magisk
1. Works without enabling systemless mode by utilizing `/magisk/.core/hosts` via Magisk's automatic `hosts` setup, as expected.
2. When enabling AdAway's systemless mode, uses `/su/etc/hosts`, as expected, and doesn't update `/magisk/.core/hosts` anymore.
3. When then disabling systemless mode, it goes back to using `/magisk/.core/hosts`, as expected.

File: AdAway/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -105,6 +105,8 @@
     <string name="download_fail_title">Download failed</string>
     <string name="no_private_file">Private file can not be created.</string>
     <string name="no_private_file_title">Private file creation failed</string>
+    <string name="disable_systemless_successful">You disabled the systemless mode.\nIt may be necessary to reboot Android for the changes to take effect.\n\nDo you want to reboot?\n(To prevent data loss make sure no app uses the sdcard at the moment!)</string>
```
Comment (voltaek) [inline AdAway/src/main/res/values/strings.xml]: "You disabled **the** systemless mode" -> "You disabled systemless mode"

File: AdAway/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -222,5 +224,9 @@
     <string name="pref_enable_debug_summary">You need to restart AdAway for this setting to take effect.</string>
     <string name="pref_enable_ipv6">Enable IPv6</string>
     <string name="pref_enable_ipv6_summary">All host file entries will get a corresponding ::1 (IPv6 localhost) entry.</string>
+    <string name="pref_enable_systemless_def">false</string>
+    <string name="pref_enable_systemless_key">enableSystemless</string>
+    <string name="pref_enable_systemless_summary"><![CDATA[Prevent AdAway to modify system partition (requires SuperSU >= 2.76)]]></string>
```
Comment (voltaek) [inline AdAway/src/main/res/values/strings.xml]: "Prevent AdAway **to modify** system partition" -> "Prevent AdAway **from modifying the** system partition"

**Comment by PerfectSlayer** [comment]
@voltaek Thanks for all your tests!

A lot a details deal about system write warning and hosts restoration. Those issue does not only apply to systemless and I think they should be treated in another issue/PR.
Concerning the alert about applicable systemless, @0-kaladin said [he wants a manual checkbox](https://github.com/AdAway/AdAway/pull/784#issuecomment-246696765) (automatic features could come later).

Anyway, it's nice to see it works great for you too! 👍

**Comment by voltaek** [comment]
@PerfectSlayer, I agree the issues I raised are probably better suited to be handled separately once we get this merged in. Especially considering this PR has already been pending since July. Hopefully @0-kaladin can review this and merge it in shortly. Worked for me as intended.

**Comment by rmenessec** [comment]
Is there a way to get a snapshot build with the fix(es)? I don't have an Android build environment.

Thanks!

**Comment by voltaek** [comment]
@rmenessec See [this comment](https://github.com/AdAway/AdAway/pull/784#issuecomment-255526820) above for a debug build. That has all the changes in this PR minus a couple interface text adjustments. I'm guessing they'll put a new build up on XDA shortly though, if they decide to bump the version for this change.

**Comment by PerfectSlayer** [comment]
@rmenessec Keep in mind it's a test  build dedicated to systemless feature. 
I had some issues to use the current build script with the latest Android Studio and I'm not sure all the native libraries are present. It's enough to test new systemless feature but some other features may be missing.

**Comment by rmenessec** [comment]
@voltaek , thanks!

@PerfectSlayer , thanks for the heads-up. I just need adblocking, automatic updates, and support for systemless SuperSU. We'll see how she goes, I guess. ;)

**Pull request merged**